lib/core/Concurrency.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/Concurrency.h>
+
+#include <core/CLogger.h>
+#include <core/CStaticThreadPool.h>
+
+#include <memory>
+#include <numeric>
+#include <thread>
+
+namespace ml {
+namespace core {
+namespace {
+//! \brief Executes a function immediately (on the calling thread).
+class CImmediateExecutor final : public CExecutor {
+public:
+    virtual void schedule(std::packaged_task<boost::any()>&& f) { f(); }
+    virtual bool busy() const { return false; }
+    virtual void busy(bool) {}
+};
+
+//! \brief Executes a function in a thread pool.
+class CThreadPoolExecutor final : public CExecutor {
+public:
+    explicit CThreadPoolExecutor(std::size_t size) : m_ThreadPool{size} {}
+
+    virtual void schedule(std::packaged_task<boost::any()>&& f) {
+        m_ThreadPool.schedule(std::forward<std::packaged_task<boost::any()>>(f));
+    }
+    virtual bool busy() const { return m_ThreadPool.busy(); }
+    virtual void busy(bool value) { return m_ThreadPool.busy(value); }
+
+private:
+    CStaticThreadPool m_ThreadPool;
+};
+
+class CExecutorHolder {
+public:
+    CExecutorHolder()
+        : m_ThreadPoolSize{0},
+    //    Work around cross compile for OS X from Linux
```

Comment:
I think this comment should be changed, because it's not the cross compilation that causes the problem but the fact that build platform for 7.x on macOS will be El Capitan, and the latest version of Xcode that works on El Capitan is 8.2.  Anyone still using Xcode 8.2 on their laptop would see the same problem.  So I'd prefer the comment to be "Workaround for Xcode 8.2 C++ headers".

Same on line 78.

Reply:
[2ec6fe8](https://github.com/elastic/ml-cpp/pull/322/commits/2ec6fe8e1b708edb10cedfd17ce47985b55f685e)